src/stan/model/indexing/assign_cl.hpp

```diff
@@ -0,0 +1,369 @@
+ * assignment
+ * @tparam ScalRhs type of the scalar on the right hand side of the assignment
+ * @param[in,out] expr_lhs expression on the left hand side of the assignment
+ * @param scal_rhs scalar on the right hand side of the assignment
+ * @param name Name of lvalue variable
+ * @param row_index index used for indexing rows of `expr_lhs`
+ * @param col_index index used for indexing columns of `expr_lhs`
+ * @throw std::out_of_range If the index is out of bounds.
+ */
+template <typename ExprLhs, typename ScalRhs,
+          require_kernel_expression_lhs_t<ExprLhs>* = nullptr,
+          require_stan_scalar_t<ScalRhs>* = nullptr>
+inline void assign(ExprLhs&& expr_lhs, const ScalRhs& scal_rhs,
+                   const char* name, const index_uni row_index,
+                   const index_uni col_index) {
+  math::block_zero_based(expr_lhs, row_index.n_ - 1, col_index.n_ - 1, 1, 1)
+      = math::constant(scal_rhs, 1, 1);
+}
+
+// assign(rev, prim, ...)
+/**
+ * Assign one primitive kernel generator expression to a reverse mode kernel
+ * generator expression, using given index.
+ *
+ * @tparam ExprLhs type of the expression on the left hand side of the
+ * assignment
+ * @tparam ExprRhs type of the expression on the right hand side of the
+ * assignment
+ * @tparam RowIndex type of index
+ * @param[in,out] expr_lhs expression on the left hand side of the assignment
+ * @param expr_rhs expression on the right hand side of the assignment
+ * @param name Name of lvalue variable
+ * @param row_index index used for indexing `expr_lhs`
+ * @throw std::out_of_range If the index is out of bounds.
+ * @throw std::invalid_argument If the right hand side size isn't the same as
+ * the indexed left hand side size.
+ */
+template <typename ExprLhs, typename ExprRhs, typename RowIndex,
+          require_rev_kernel_expression_t<ExprLhs>* = nullptr,
+          require_all_kernel_expressions_and_none_scalar_t<ExprRhs>* = nullptr>
+inline void assign(ExprLhs&& expr_lhs, const ExprRhs& expr_rhs,
+                   const char* name, const RowIndex& row_index) {
+  if (std::is_same<RowIndex, index_omni>::value) {
+    stan::math::check_size_match("omni assign", "left hand side rows",
+                                 expr_lhs.rows(), name, expr_rhs.rows());
+    stan::math::check_size_match("omni assign", "left hand side columns",
+                                 expr_lhs.cols(), name, expr_rhs.cols());
+  }
+  auto&& lhs_holder = rvalue(expr_lhs.val_op(), name, row_index);
+  auto& lhs = lhs_holder;
+  math::arena_matrix_cl<double> prev_vals = lhs;
+  lhs = expr_rhs;
+  math::reverse_pass_callback([expr_lhs, row_index, prev_vals, name]() mutable {
+    auto&& lhs_val = rvalue(expr_lhs.val_op(), name, row_index);
+    auto&& lhs_adj = rvalue(expr_lhs.adj(), name, row_index);
+    math::results(lhs_val, lhs_adj) = math::expressions(
+        prev_vals, math::constant(0.0, lhs_adj.rows(), lhs_adj.cols()));
+  });
+}
```

Comment:
I'm confused about what's going on here, where is the `expr_rhs` getting assigned to the `expr_lhs`'s values? Is it because `lhs` is a ref to `lhs_holder` and `rvalue()` returns a reference to the `expr_lhs` values? So then `lhs = expr_rhs` is actually assigning to `expr_lhs`'s values?

Reply:
Exactly.

Reply:
Can you leave a note on the docs on what's going on here? It's a bit hard to read

Reply:
Added some comments.